File: Trees_and_Graphs/minHeap.js
class MinHeap {

    constructor() {
        this.heap = [null]
    }
    add(val) {
        this.heap.push(val)
        if (this.heap.length > 2)
            this.heapify(this.heap.length - 1)
    }
    pop() {
        if (this.heap.length < 2) return;
        else if(this.heap.length == 2){
            const value = this.heap.pop()
            return value
        } else if (this.heap.length == 3){
            const smallest = this.heap[0]
            this.heap[0] = this.heap.pop()
            return smallest
        }
         else {
            const smallest = this.heap[0]
            this.heap[0] = this.heap.pop()
            this.heapifyDown()
            return smallest
         }
        if (this.heap.length > 2)
            this.heapifyUp(this.heap.length - 1)
    }
    heapifyUp(index) {
        if (index <= 1) return;
        let parent = Math.floor(index / 2)
        if (this.heap[parent] > this.heap[index]) {
            [this.heap[parent], this.heap[index]] = [this.heap[index], this.heap[parent]]
            this.heapifyUp(parent)
        }
    }

    heapifyDown(index = 1){
        const parent = Math.floor(index /2)
        const left = Math.floor(index /2)

    }
}

let minHeap = new MinHeap()
minHeap.add(5)
minHeap.add(4)
minHeap.add(6)
minHeap.add(2)
minHeap.add(11)
console.log(minHeap.heap)

File: Trees_and_Graphs/python/minHeap.py
from math import floor, isinf


class MinHeap:
    def __init__(self):
        self.heap = [float('-inf')]

    def add(self, val):
        self.heap.append(val)
        index = len(self.heap) - 1
        self.heapifyUp(index)

    def heapifyUp(self, index: int):
        parent = floor(index / 2)

        if self.heap[index] <= 1:
            return

        if self.heap[parent] > self.heap[index]:
            temp = self.heap[parent]
            self.heap[parent] = self.heap[index]
            self.heap[index] = temp
            self.heapifyUp(parent)

    def pop(self):
        if len(self.heap) is 1:
            return
        small = self.heap[1]
        if len(self.heap) < 3:
            return self.heap.pop()
        self.heap[1] = self.heap.pop()
        if len(self.heap) < 4:
            return small
        else:
            self.heapifyDown(1)
        return small

    def heapifyDown(self, parent: int):
        left = (parent * 2)
        right = left + 1
        leftChild = self.indexCatch(left) or float('inf')
        rightChild = self.indexCatch(right) or float('inf')

        if isinf(leftChild) and isinf(rightChild):
            return

        if leftChild < rightChild:
            if self.heap[parent] > self.heap[left]:
                temp = self.heap[parent]
                self.heap[parent] = self.heap[left]
                self.heap[left] = temp
                self.heapifyDown(left)
        else:
            if self.heap[parent] > self.heap[right]:
                temp = self.heap[parent]
                self.heap[parent] = self.heap[right]
                self.heap[right] = temp
                self.heapifyDown(right)

    def indexCatch(self, index):
        try:
            return self.heap[index]
        except Exception as e:
            return None


minHeap = MinHeap()
minHeap.add(5)
minHeap.add(4)
minHeap.add(6)
minHeap.add(2)
minHeap.add(11)
print(minHeap.heap)
minHeap.pop()
print(minHeap.heap)
minHeap.pop()
print(minHeap.heap)
minHeap.pop()
print(minHeap.heap)
minHeap.pop()
print(minHeap.heap)
minHeap.pop()
print(minHeap.heap)
minHeap.pop()
print(minHeap.heap)
